source/agora/common/TransactionPool.d

```diff
@@ -128,26 +128,35 @@ public class TransactionPool
 
     ***************************************************************************/
 
-    public bool add (Transaction tx) @safe
+    public bool add (Transaction tx) @safe nothrow
     {
-        static ubyte[] buffer;
+        try
+        {
+            static ubyte[] buffer;
 
-        // check double-spend
-        if (!isValidTransaction(tx))
-            return false;
+            // check double-spend
+            if (!isValidTransaction(tx))
+                return false;
 
-        // insert each input information of the transaction
-        foreach (input; tx.inputs)
-            this.input_set.put(input.hashFull());
+            // insert each input information of the transaction
+            foreach (const ref input; tx.inputs)
+                this.input_set.put(input.hashFull());
 
-        serializeToBuffer(tx, buffer);
+            serializeToBuffer(tx, buffer);
 
-        () @trusted {
-            db.execute("INSERT INTO tx_pool (key, val) VALUES (?, ?)",
-                hashFull(tx)[], buffer);
-        }();
+            () @trusted {
+                db.execute("INSERT INTO tx_pool (key, val) VALUES (?, ?)",
+                    hashFull(tx)[], buffer);
+            }();
 
-        return true;
+            return true;
+        }
+        catch (Exception ex)
+        {
+            log.error("Error adding transaction: {}" ~
+                "Transaction: {}", ex.msg, tx);
+            return false;
+        }
```

Comment:
Can the whole function throws ? It looks like at least what's before the `foreach` cannot.